tests/codegen-llvm/autodiff/abi_handling.rs

```diff
@@ -0,0 +1,210 @@
+//@ revisions: debug release
+
+//@[debug] compile-flags: -Zautodiff=Enable -C opt-level=0 -Clto=fat
+//@[release] compile-flags: -Zautodiff=Enable -C opt-level=3 -Clto=fat
+//@ no-prefer-dynamic
+//@ needs-enzyme
+
+// This test checks that Rust types are lowered to LLVM-IR types in a way
+// we expect and Enzyme can handle. We explicitly check release mode to
+// ensure that LLVM's O3 pipeline doesn't rewrite function signatures
+// into forms that Enzyme can't process correctly.
+
+#![feature(autodiff)]
+
+use std::autodiff::{autodiff_forward, autodiff_reverse};
+
+#[derive(Copy, Clone)]
+struct Input {
+    x: f32,
+    y: f32,
+}
+
+#[derive(Copy, Clone)]
+struct Wrapper {
+    z: f32,
+}
+
+#[derive(Copy, Clone)]
+struct NestedInput {
+    x: f32,
+    y: Wrapper,
+}
+
+fn square(x: f32) -> f32 {
+    x * x
+}
+
+// CHECK: ; abi_handling::df1
+// CHECK-NEXT: Function Attrs
+// debug-NEXT: define internal { float, float } @_ZN12abi_handling3df117h55144f697f16c530E
+// debug-SAME: (ptr align 4 %x, ptr align 4 %bx_0)
+// release-NEXT: define internal fastcc float @_ZN12abi_handling3df117h55144f697f16c530E
+// release-SAME: (float %x.0.val, float %x.4.val)
+
+// CHECK: ; abi_handling::f1
+// CHECK-NEXT: Function Attrs
+// debug-NEXT: define internal float @_ZN12abi_handling2f117hd2edd01111f953c8E
+// debug-SAME: (ptr align 4 %x)
+// release-NEXT: define internal fastcc noundef float @_ZN12abi_handling2f117hd2edd01111f953c8E
```

Comment:
I am worried that the tests could fail in the future because of mangling name updates. At the same time, I remember us discussing `no_mangle` having an effect, because it maces a function publicly visible and hence limiting what LLVM will optimize. I would probably just not match the mangled name itself, unless you also see matches on mangled names in other tests?

Also, could you use CHECK-LABEL for the matching of the debug info name like `abi_handling::df2`? I think that suitable here and e.g. the consts.rs file uses it.
https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-label-directive